### PR TITLE
feat(unstable): Support data: urls

### LINKED ISSUE #5059: Support importing data URLs
As pointed out in #5048, we should consider importing data URLs.  The following works in Chromium:

```js
await import("data:application/javascript;charset=utf-8;base64,Y29uc29sZS5sb2coImhlbGxvIHdvcmxkIik7");
```

Which which then will log `hello world` to the console.

### PR DESCRIPTION
**Pull request by SyrupThinker**
Closes #5059 

Notes:
- `percent-encoding` is already an indirect dependency of `url`
- The disk cache name might need revising
- Graceful failure is probably desired in `extract_data_url`, what kind of Error should be returned?

### CODE REVIEW

**Comment by SyrupThinker** [comment]
> What happens with media types we don't support, like `text/plain`? It would be good to have some non-happy path tests for this.

Currently they get treated as JavaScript because it falls through here:
https://github.com/denoland/deno/blob/fca204d8335b407236e16e639099a7696f9563ee/cli/global_state.rs#L146-L149

**Comment by kitsonk** [comment]
Hrmmm... I didn't think we allowed media types we didn't know.  @bartlomieju are we sure we want that?  Seems like a big security hole.

Non-happy path tests would still be good.

**Comment by SyrupThinker** [comment]
> Non-happy path tests would still be good.

Working on that and checking whether removing that fallback causes regressions

**Comment by SyrupThinker** [comment]
I'd like to have some feedback on how to handle `extract_data_url` and unknown media type failure first because the non-happy test results depend on that.

File: cli/tests/integration_tests.rs [hunk 1]
```diff
@@ -572,6 +572,21 @@ fn bundle_import_map() {
   assert_eq!(output.stderr, b"");
 }
 
+#[test]
+fn data_import() {
+  let status = util::deno_cmd()
+    .current_dir(util::tests_path())
+    .arg("test")
+    .arg("--unstable")
```
Comment (ry) [inline cli/tests/integration_tests.rs:580]: Does it actually need --unstable? It seems like it would work without.

Comment (SyrupThinker) [reply cli/tests/integration_tests.rs:580]: Ah, a bad case of copy pasta

**Comment by SyrupThinker** [comment]
> How are data-urls used with javascript typically?
The usages I encounter are generally data embedding.

> Do data urls work like this in Chrome?
Both Firefox and Chrome allow imports of that form.
This implementation diverges in behaviour when the data url is malformed or has an incompatible media type.
The browser throw a type error, this currently panics for the former and treats it as JavaScript in the latter.

**Comment by ry** [comment]
Ok sounds good. 

I think we should make this an unstable feature for the time being. I'd leave it up to you to figure out how to do that exactly (if you don't mind). Here's reference:
https://github.com/denoland/deno/blob/2b02535028f868ea8dfc471c4921a237747ccd4a/cli/state.rs#L236-L248

**Comment by bartlomieju** [comment]
Thanks for the patch @SyrupThinker; however I think we should wait on this PR after 1.0 is released and compiler refactor is finished. 

It introduces yet another layer to consider and if I'm not mistaken data imports are dependent on location/origin which we currently don't support.

> Hrmmm... I didn't think we allowed media types we didn't know. @bartlomieju are we sure we want that? Seems like a big security hole.

That's how it always worked: unknown media types are treated as if they were JS which in most cases ends up in an error from V8 or TS compiler.

**Comment by SyrupThinker** [comment]
Ok, I'll continue working in this after the redactor then.

**Comment by SyrupThinker** [comment]
> It introduces yet another layer to consider and if I'm not mistaken data imports are dependent on location/origin which we currently don't support.

That is correct; we certainly don't want to allow circumventing restrictions by imports like `https > data > file` or `https > data > http`.

**Comment by ry** [comment]
@SyrupThinker I'm sorry we didn't get this landed earlier. If you can get this rebased on master, I would include it in 1.4.0 (to be released September 13)

**Comment by ry** [comment]
Since the main use case for this feature is for loading workers, it would be good to have a test case demonstrating that you can start a worker with a data url.

**Comment by SyrupThinker** [comment]
I don't know how to resolve the tsc errors (or how to debug them), could use some help for that.

For example I can't figure out how to log stuff in `cli/tsc/00_typescript.js`.

File: cli/tests/integration_tests.rs [hunk 2]
```diff
@@ -950,6 +950,46 @@ fn bundle_import_map() {
 }
 
 #[test]
+fn data_import() {
+  let status = util::deno_cmd()
+    .current_dir(util::tests_path())
+    .arg("test")
+    .arg("--unstable")
```
Comment (bartlomieju) [inline cli/tests/integration_tests.rs:957]: @SyrupThinker please rewrite those tests as `itest!`

File: cli/tsc.rs [hunk 3]
```diff
@@ -687,6 +687,7 @@ impl TsCompiler {
       self.file_fetcher.clone(),
       global_state.maybe_import_map.clone(),
       permissions.clone(),
+      global_state.flags.unstable,
```
Comment (bartlomieju) [inline cli/tsc.rs:690]: Not sure if we need to hide it behind unstable

**Review by bartlomieju — COMMENTED** [review]
[MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs) tells that `data:` URLs are not allowed at top level due to phishing vulnerability. I wonder if we also should disallow them as entry point 🤔

**Comment by bartlomieju** [comment]
@SyrupThinker I have fixed the tests. TS failures were caused by #7392

**Review by bartlomieju — APPROVED** [review]
LGTM, thank you @SyrupThinker

**Pull request merged**